### PR TITLE
Add filtrable table

### LINKED ISSUE #145: Create filterable table component
Sortable table component will be used on most daashboard pages:
https://www.figma.com/file/rdg8gZtSWNYaVtupMtdhST/CoVid---Stam-Acasa?node-id=145%3A16

**Default state**
Screen Shot 2020-04-06 at 12 31 33 PM

**Clicked state**
Screen Shot 2020-04-06 at 12 30 58 PM

### PR DESCRIPTION
**Pull request by maloun96**
Fixed #145 
Create filtered table, split DashboardWidget

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -20,7 +20,7 @@
     "react-super-responsive-table": "^5.1.1"
   },
   "scripts": {
-    "start": "./env.sh && react-scripts start",
+    "start": "env.sh && react-scripts start",
```
Comment (surdu) [inline frontend/package.json]: This looked correct initially

File: frontend/src/components/BaseDashboard/components/DashboardWidgets/index.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import DashboardWidget from "../DashboardWidget";
+import React from "react";
+
+const DashboardWidgets = () => (
+  <>
+    <strong>Date colectate</strong>
+    <br />
```
Comment (surdu) [inline frontend/src/components/BaseDashboard/components/DashboardWidgets/index.js]: Please use CSS margin instead of a bunch of`<br />`

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  };
+
+  return (
+    <div className="filterColumn" ref={ref}>
+      <span>{name}</span>{" "}
+      <button
+        className="filter-buttom"
+        onClick={() => setIsVisible(!isVisible)}
+      >
+        <img src={require("../../../../images/filter-icon.svg")} alt="" />
```
Comment (surdu) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: There is a [nicer way](https://github.com/code4romania/stam-acasa/blob/develop/frontend/src/components/Header/Logo.js#L7) to achieve this

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import * as React from "react";
+import { useState } from "react";
+
+const useTableFilter = items => {
+  const copy = { ...items[0] };
+  Object.keys(copy).map(key => (copy[key] = []));
```
Comment (surdu) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]:
```suggestion
  Object.keys(copy).forEach(key => (copy[key] = []));
```

**Review by surdu — CHANGES_REQUESTED** [review]

File: frontend/src/components/BaseDashboard/components/DashboardWidgets/index.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    <br />
+    <div className="columns is-desktop">
+      <div className="column is-one-third-desktop">
+        <div className="dashboard-box" style={{ padding: 0 }}>
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/DashboardWidgets/index.js]: don't use style

File: frontend/src/components/BaseDashboard/components/DashboardWidgets/index.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          <DashboardWidget
+            title={"Utilizatori red flag in ultimile 24 de ore"}
+            data={[100, 200, 300, 200]}
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/DashboardWidgets/index.js]: you should put those numbers in a constant (they look random at the moment - even more so in a week from now)

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  const [isVisible, setIsVisible] = useState(false);
+  const [searchTerm, setSearchTerm] = useState("");
+  const [searchResults, setSearchResults] = useState([]);
+  const ref = useRef();
+  values = values = values.filter((x, i, a) => a.indexOf(x) === i);
+
+  useOutsideClick(ref, () => setIsVisible(false));
+
+  useEffect(() => {
+    setSearchResults((values = values.filter((x, i, a) => a.indexOf(x) === i)));
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: don't assign to props

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import PropTypes from "prop-types";
+import "./FilterColumn.scss";
+import useOutsideClick from "../../../useOutsideClick";
+
+const FilterColumn = ({ name, values, handleSelect, selectedValues }) => {
+  const [isVisible, setIsVisible] = useState(false);
+  const [searchTerm, setSearchTerm] = useState("");
+  const [searchResults, setSearchResults] = useState([]);
+  const ref = useRef();
+  values = values = values.filter((x, i, a) => a.indexOf(x) === i);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: don't assign to props

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js (hunk 7 above)
Comment (tudorgergely) [reply frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: this also basically sets searchResults to undefined (you are assigning an assignment )

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        onClick={() => setIsVisible(!isVisible)}
+      >
+        <img src={require("../../../../images/filter-icon.svg")} alt="" />
+      </button>
+      <div className={["searchBox", isVisible ? "isVisible" : ""].join(" ")}>
+        <input type="text" placeholder={name} onChange={handleChange} />
+        <ul>
+          {searchResults.length > 0
+            ? searchResults.map((value, index) => (
+                <li key={index} onClick={() => handleSelect(value)}>
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: don't use index as key https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+                  {value}
+                  {selectedValues.includes(value) && <span>&#10003;</span>}
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: use css content + after pseudoselector for this

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                </li>
+              ))
+            : "Nici un rezultat"}
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: Niciun

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  const [filters, setFilters] = useState(copy);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]: this won't work if copy changes (if items changes). Replace with useEffect

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
+import * as React from "react";
+import { useState } from "react";
+
+const useTableFilter = items => {
+  const copy = { ...items[0] };
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]: all the items have the same fields? can we count on the first item keys?

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+  const filteredItems = React.useMemo(() => {
+    return items.filter(item => {
+      for (var key in filters) {
+        if (filters[key].length > 0 && !filters[key].includes(item[key])) {
+          return false;
+        }
+      }
+      return true;
+    });
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]:
```suggestion
    return items.filter(item => Object.keys(filters).every(key => filters[key].length && filters[key].includes(item[key]));
```

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  const addOrRemove = (array, value) => {
+    var index = array.indexOf(value);
+
+    if (index === -1) {
+      array.push(value);
+    } else {
+      array.splice(index, 1);
+    }
+
+    return array;
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]: try not to change the params of your function (no push/splice)

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  };
+
+  const requestFilter = (column, value) => {
+    if (!filters[column]) {
+      filters[column] = [];
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]: use setFilters, don't assign directly

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    filters[column] = addOrRemove(filters[column], value);
+    setFilters({ ...filters });
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]: never assign to you state, use setFilters everywhere

**Review by tudorgergely — CHANGES_REQUESTED** [review]

File: frontend/src/components/BaseDashboard/pages/Reports/index.js [hunk 18]
```diff
@@ -0,0 +1,124 @@
+import React from "react";
+import { Table, Thead, Tbody, Tr, Th, Td } from "react-super-responsive-table";
+import "react-super-responsive-table/dist/SuperResponsiveTableStyle.css";
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/pages/Reports/index.js:3]: why not react-table? I don't mind this lib (although it's not very popular), but I'm curious

File: frontend/src/components/useOutsideClick/index.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+    };
+
+    document.addEventListener("mousedown", listener);
+    document.addEventListener("touchstart", listener);
+
+    return () => {
+      document.removeEventListener("mousedown", listener);
+      document.removeEventListener("touchstart", listener);
+    };
+  }, [ref, handler]);
```
Comment (tudorgergely) [inline frontend/src/components/useOutsideClick/index.js:19]: Listening for ref changes doesn't really work: https://reactjs.org/docs/hooks-reference.html#useref

```
Keep in mind that useRef doesn’t notify you when its content changes. 
Mutating the .current property doesn’t cause a re-render. 
If you want to run some code when React attaches or detaches a ref 
to a DOM node, you may want to use a callback ref instead.
```

You should try and use e ref callback instead (let me know if you need some help)

File: frontend/src/components/BaseDashboard/pages/Reports/index.js (hunk 18 above)
Comment (maloun96) [reply frontend/src/components/BaseDashboard/pages/Reports/index.js:3]: because it's responsive, try to resize window to mobile width @tudorgergely

File: frontend/src/components/useOutsideClick/index.js (hunk 19 above)
Comment (maloun96) [reply frontend/src/components/useOutsideClick/index.js:19]: and yes, i have a callback there.
```const listener = event => {
      if (ref.current.contains(event.target)) {
        return;
      }
      handler(event);
    };```

https://medium.com/@kevinfelisilda/click-outside-element-event-using-react-hooks-2c540814b661
@tudorgergely

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+import "./FilterColumn.scss";
+import useOutsideClick from "../../../useOutsideClick";
+import { ReactComponent as FilterSvg } from "../../../../images/filter-icon.svg";
+
+const FilterColumn = ({ name, values, handleSelect, selectedValues }) => {
+  const [isVisible, setIsVisible] = useState(false);
+  const [searchTerm, setSearchTerm] = useState("");
+  const [searchResults, setSearchResults] = useState([]);
+  const ref = useRef();
+  const uniqueValues = values.filter((x, i, a) => a.indexOf(x) === i);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: why is this needed if we have searchResults?

**Review by tudorgergely — CHANGES_REQUESTED** [review]

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js (hunk 20 above)
Comment (maloun96) [reply frontend/src/components/BaseDashboard/components/FilterColumn/index.js]:
```
setSearchResults(
      uniqueValues.filter(item =>
        item.toLowerCase().includes(searchTerm.toLowerCase())
      )
    );
```

when user search second time it should search throw all values not only in searchResult

Comment (tudorgergely) [reply frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: I see, you can do the filter inside that set, no need for a separate variable (performance might be better since now you're filtering on every render)

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+
+  useOutsideClick(ref, () => setIsVisible(false));
+
+  useEffect(() => {
+    setSearchResults(uniqueValues);
+  }, []);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: you need to add uniqueValues here or it might have old values

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+  useEffect(() => {
+    setSearchResults(
+      uniqueValues.filter(item =>
+        item.toLowerCase().includes(searchTerm.toLowerCase())
+      )
+    );
+  }, [searchTerm]);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]: need to add uniqueValues here as well

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 23]
```diff
@@ -0,0 +1,44 @@
+import * as React from "react";
+import { useState } from "react";
+import { useEffect } from "react";
+
+const useTableFilter = items => {
+  const copy = { ...items[0] };
+  Object.keys(copy).forEach(key => (copy[key] = []));
+  const [filters, setFilters] = useState([]);
+
+  useEffect(() => setFilters(copy), []);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]: copy needs to go in the deps array

**Review by tudorgergely — COMMENTED** [review]
a few more things and a rebase @maloun96

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 24]
```diff
@@ -0,0 +1,41 @@
+import * as React from "react";
+import { useState } from "react";
+import { useEffect } from "react";
+
+const useTableFilter = items => {
+  const copy = { ...items[0] };
+  Object.keys(copy).forEach(key => (copy[key] = []));
+  const [filters, setFilters] = useState([]);
+
+  useEffect(() => setFilters(copy), [items[0]]);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]:
```suggestion
  const [filters, setFilters] = useState({});

  useEffect(() => setFilters(
      Object.keys(items[0]).reduce((acc, key) => ({ ...acc, [key]: []}))
  ), [items[0]]);
```

File: frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+    if (index === -1) {
+      return [...array, value];
+    } else {
+      return array.filter(el => el !== value);
+    }
+  };
+
+  const requestFilter = (column, value) => {
+    let selectedFilter = filters[column] ? filters[column] : [];
+    selectedFilter = addOrRemove(selectedFilter, value);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/use-table-filter.js]:
```suggestion
    const selectedFilter = addOrRemove(filters[column] || [], value);
```

File: frontend/src/components/BaseDashboard/components/FilterColumn/index.js [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+
+  useOutsideClick(ref, () => setIsVisible(false));
+
+  useEffect(() => {
+    setSearchResults(
+      uniqueValues.filter(item =>
+        item.toLowerCase().includes(searchTerm.toLowerCase())
+      )
+    );
+  }, [searchTerm, values]);
```
Comment (tudorgergely) [inline frontend/src/components/BaseDashboard/components/FilterColumn/index.js]:
```suggestion
  useOutsideClick(ref, () => setIsVisible(false));

  useEffect(() => {
    setSearchResults(
      values
        .filter(item => item.toLowerCase().includes(searchTerm.toLowerCase()))
        .filter((x, i, a) => a.indexOf(x) === i)
    );
  }, [searchTerm, values]);
```

**Review by tudorgergely — APPROVED** [review]

**Review by surdu — APPROVED** [review]
